src/shogun/lib/basetag.h

```diff
@@ -0,0 +1,138 @@
+ *     this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: 2016, Sergey Lisitsyn <lisitsyn.s.o@gmail.com>
+ *          2016, Sanuj Sharma <sanuj.sharma.in@gmail.com>
+ */
+
+#ifndef _BASETAG_H_
+#define _BASETAG_H_
+
+#include <string>
+
+namespace shogun
+{
+    /** @brief Parent class of Tag
+     *
+     * Stores name and hash for a Tag.
+     */
+    class BaseTag
+    {
+    public:
+
+        /** Constructor to initialize hash from name
+         * @param _name name for tag
+         */
+        explicit BaseTag(const std::string& _name) : 
+            m_name(_name), m_hash(std::hash<std::string>()(_name))
+        {
+        }
+
+        /** Copy constructor
+         * @param other basetag object to be copied
+         */
+        BaseTag(const BaseTag& other) : 
+            m_name(other.m_name), m_hash(other.m_hash)
+        {
+        }
+
+        /** Class Assignment operator
+         * @param other basetag object to be assigned
+         */
+        BaseTag& operator=(const BaseTag& other)
+        {
+            m_name = other.m_name;
+            m_hash = other.m_hash;
+            return *this;
+        }
+        
+        /** Returns name of Tag
+         * @return name of basetag
+         */
+        inline std::string name() const
+        {
+            return m_name;
+        }
+
+        /** Returns hash of Tag
+         * @return hash value
+         */
+        inline std::size_t hash() const
+        {
+            return m_hash;
+        }
+
+        /** Equality operator
+         * @param first first BaseTag
+         * @param second secondBaseTag
+         */
+        friend inline bool operator==(const BaseTag& first, const BaseTag& second);
+
+        /** Inequality operator
+         * @param first first BaseTag
+         * @param second secondBaseTag
+         */
+        friend inline bool operator!=(const BaseTag& first, const BaseTag& second);
+
+    private:
+        std::string m_name;
+        size_t m_hash;
+    };
+
+    inline bool operator==(const BaseTag& first, const BaseTag& second)
+    {
+        return first.m_hash == second.m_hash ? first.m_name == second.m_name : false;
```

Comment:
I dont understand how the name can be different when the hash is the name.
Is it because of collision?